File: c++/leetcode/290. Word Pattern.cpp
class Solution
{
public:
  bool wordPattern(string pattern, string str)
  {
    vector<string> words;
    string tmp = "";
    for (auto i : str)
    {
      if (i == ' ')
      {
        words.push_back(tmp);
        tmp = "";
      }
      else
        tmp += i;
    }

    words.push_back(tmp);

    if (pattern.size() != words.size())
      return false;

    int i = 0;
    unordered_map<char, string> seen;
    unordered_set<string> dupe;

    for (auto j : pattern)
    {
      auto it = seen.find(j);

      if (it != seen.end())
      {
        if (it->second != words[i])
          return false;
      }
      else
      {
        if (dupe.count(words[i]))
          return false;

        seen.insert({j, words[i]});
        dupe.insert(words[i]);
      }

      ++i;
    }

    return true;
  }
};

File: javascript/leetcode/290. Word Pattern.js
/**
 * @param {string} pattern
 * @param {string} str
 * @return {boolean}
 */
var wordPattern = function (pattern, str) {
  const words = str.split(" ");

  if (words.length !== pattern.length) return false;

  const seen = new Map();
  const dupe = new Set();

  let i = 0;
  for (let j of pattern) {
    if (seen.has(j)) {
      if (seen.get(j) !== words[i]) return false;
    } else {
      if (dupe.has(words[i])) return false;
      seen.set(j, words[i]);
      dupe.add(words[i]);
    }

    ++i;
  }

  return true;
};
